suse_migration_services/units/product_setup.py

```diff
@@ -66,53 +68,66 @@ def main():
                     '/tmp/products.d.backup/'
                 ]
             )
-        baseproduct_filename = get_baseproduct(products_metadata)
-        if len(baseproduct_filename) == 1:
-            baseproduct = baseproduct_filename.pop()
-        else:
-            if not baseproduct_filename:
-                log.error('There is no baseproduct')
-                raise DistMigrationProductSetupException(
-                    'There is no baseproduct'
-                )
+        baseproduct = get_baseproduct(products_metadata)
+        if baseproduct.count != 1:
+            if not baseproduct.count:
+                message = 'There is no baseproduct'
+                log.error(message)
+                raise DistMigrationProductSetupException(message)
             else:
-                log.error('There are more than 1 product without flavor tag')
-                raise DistMigrationProductSetupException(
-                    'There are more than 1 product without flavor tag'
-                )
+                message = ('Found multiple product definitions'
+                           'without element <flavor>')
+                log.error(message)
+                raise DistMigrationProductSetupException(message)
         log.info('Updating Base Product to be suitable for migration')
-        update_baseproduct(baseproduct)
+        update_baseproduct(baseproduct.prod_filename)
     except Exception as issue:
         message = 'Base Product update failed with {0}'.format(issue)
         log.error(message)
         raise DistMigrationProductSetupException(message)
 
 
 def get_baseproduct(products_metadata):
-    baseproduct_filename = []
-    xml = ElementTree()
-    for product_name in os.listdir(products_metadata):
-        product = os.sep.join(
-            [products_metadata, product_name]
-        )
-        xml.parse(product)
-        register_sections = xml.findall('register')
-        for register in register_sections:
-            flavor = register.findall('flavor')
-            if not flavor:
-                baseproduct_filename.append(product)
+    base_product_type = namedtuple(
+        'base_product_type', ['count', 'prod_filename']
+    )
+    prod_filenames = glob.glob(
+        os.path.join(products_metadata, '*.prod')
+    )
+    count = 0
+    name = None
+    try:
+        xml = ElementTree()
+        for prod_filename in prod_filenames:
+            xml.parse(prod_filename)
+            register_sections = xml.findall('register')
+            for register in register_sections:
+                flavor = register.findall('flavor')
+                if not flavor:
+                    count = count + 1
+                    name = prod_filename
 
-    return baseproduct_filename
+        return base_product_type(count, name)
```

Comment:
While we count the number of files that do not have the <flavor> element, this construct looses the information about which files have the missing element, i.e. you may end up with count having the value 3, and name has the value of the last file that did not have the <flavor> element. This is confusing. We either only count and then complain later if count > 1 as is currently implemented, or we count and keep the filename information, but then we need to keep track of all the filenames that have the problem. The latter approach makes for better error reporting and thus we should go with that.

Reply:
I agree on that, we should have the proper filenames available with the counter

Reply:
I agree that the more info we show, the better. I understood, in the previous comments, if count > 1 , prod_name would be None so I did not keep the names. I will adapt the code to reflect the changes and keeping the names.